compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs

```diff
@@ -178,7 +178,7 @@ impl Binary {
                     let non_constant = if lhs.is_some() { self.rhs } else { self.lhs };
                     if let Some(constant) = constant {
                         let max_possible_value =
-                            2u128.pow(dfg.get_value_max_num_bits(non_constant)) - 1;
+                            2u128.pow(dfg.type_of_value(non_constant).bit_size()) - 1;
```

Comment:
There is actually two bugs. The `Cast` logic was incorrect and also that in https://github.com/noir-lang/noir/pull/4691 we should be checking against the bit_size of variable's type. According to the comments for `get_value_max_num_bits` it is attempting to `Returns the maximum possible number of bits that `value` can potentially be`.

Reply:
I'll pull this out into its own fix and we can wait for @TomAFrench to take a look at the `Cast` fix as those are his changes.

Reply:
Can you explain why we should make this change? The motivation for the current implementation is that `get_value_max_num_bits` will return the max number of bits which `value` can take, i.e. `non_constant < 2^max_bits`. This allows us to reason how about when the constant is greater than this upper bound then it's impossible for the equality to hold.

This relies on casts up from a smaller type as otherwise the constant will always be less than the max potential value of the variable.

Reply:
> This relies on casts up from a smaller type as otherwise the constant will always be less than the max potential value of the variable.

Ok I understand thank you for elaborating. I was thinking that using `get_value_max_num_bits` was a typo due to the comments above about variable types. I can update those to discuss that we want to check how for casts from a lower type as well as bringing usage back of `get_value_max_num_bits`.